### PR TITLE
Log safe implemented

### PR DESCRIPTION
**Pull request by tillyow**
## General
**Before this PR**:
We did not implement Safe Logging via the AtlasDbDependencyException.

**After this PR**:

==COMMIT_MSG==
AtlasDbDependencyException implements SafeLoggable.
==COMMIT_MSG==

**Priority**:
P2
**Concerns / possible downsides (what feedback would you like?)**:
No
**Is documentation needed?**:
No
## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:
No
**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:
No
**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:
No
**Does this PR need a schema migration?**
No
## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:

**What was existing testing like? What have you done to improve it?**:

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:

**Has the safety of all log arguments been decided correctly?**:

**Will this change significantly affect our spending on metrics or logs?**:

**How would I tell that this PR does not work in production? (monitors, etc.)**:

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:

## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@raiju

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientFactory.java [hunk 1]
```diff
@@ -99,12 +101,18 @@ public CassandraClient create() {
...
             return instrumentClient(getRawClientWithKeyspaceSet());
         } catch (Exception e) {
-            String message = String.format(
-                    "Failed to construct client for %s/%s", cassandraServer.proxy(), clientConfig.keyspace());
             if (clientConfig.usingSsl()) {
-                message += " over SSL";
+                throw new ClientCreationFailedException(
+                        "Failed to construct client for {}/{} over SSL",
+                        e,
+                        UnsafeArg.of("cassandraServer.proxy", cassandraServer.proxy()),
```
Comment (mdaudali) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientFactory.java]: cassandraServerProxy or proxy is sufficient! We don't need to name the arg the same as the variable name.

Also, what led you to it being unsafe? I haven't checked yet, just want to check your reasoning.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientFactory.java [hunk 2]
```diff
@@ -99,12 +101,18 @@ public CassandraClient create() {
...
+                        SafeArg.of("clientConfig.keyspace", clientConfig.keyspace()));
```
Comment (mdaudali) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientFactory.java]: Same as above - keyspace is fine.

File: atlasdb-commons/src/main/java/com/palantir/common/exception/AtlasDbDependencyException.java [hunk 3]
```diff
@@ -15,18 +15,61 @@
...
+import javax.annotation.Nullable;
 
-    public AtlasDbDependencyException(String msg) {
-        super(msg);
+public class AtlasDbDependencyException extends RuntimeException implements SafeLoggable {
+    private static final String ATLASDB_DEPENDENCY_LOG_MESSAGE = "AtlasDB dependency threw an exception.";
+    private final List<Arg<?>> args;
+    private final String logMessage;
+
+    public AtlasDbDependencyException(@CompileTimeConstant final String logMessage) {
```
Comment (mdaudali) [inline atlasdb-commons/src/main/java/com/palantir/common/exception/AtlasDbDependencyException.java]: nit: String doesn't need to be marked final here. We tend not to mark variables as final - there's an internal style guide floating around that's worth reading up!

Comment (mdaudali) [reply atlasdb-commons/src/main/java/com/palantir/common/exception/AtlasDbDependencyException.java]: also nit: Don't forget the advice I gave in the call! You don't need one constructor with args and one without, and you should just have the one private constructor at the end setting the fields. See KeyAlreadyExistsException as an example.

**Review by mdaudali — COMMENTED** [review]
Almost there!

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientFactory.java (hunk 1 above)
Comment (tillyow) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientFactory.java]: It threw a compiletime error saying it was unsafe

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientFactory.java [hunk 4]
```diff
@@ -105,12 +107,18 @@ public CassandraClient create() {
...
         } catch (Exception e) {
-            String message = String.format(
-                    "Failed to construct client for %s/%s", cassandraServer.proxy(), clientConfig.keyspace());
             if (clientConfig.usingSsl()) {
-                message += " over SSL";
+                throw new ClientCreationFailedException(
+                        "Failed to construct client for {}/{} over SSL",
+                        e,
+                        SafeArg.of("proxy", CassandraLogHelper.host(cassandraServer.proxy())),
+                        UnsafeArg.of("Keyspace", clientConfig.keyspace()));
```
Comment (mdaudali) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientFactory.java]: nit: lowercase K :) 

Also, we actually do log keyspace as safe elsewhere (and unsafe too...). Discussing internally, we believe this actually can be made safe here (and the example you found).

**Review by mdaudali — APPROVED** [review]

**Pull request merged**